pkg/controller/composite/hooks.go

```diff
@@ -39,7 +41,14 @@ func (pc *parentController) callHook(
 	// First check if we should instead call the finalize hook,
 	// which has the same API as the sync hook except that it's
 	// called while the object is pending deletion.
-	if request.Parent.GetDeletionTimestamp() != nil && pc.finalizeHook.IsEnabled() {
+	//
+	// In addition to finalizing when the object is deleted, we also finalize
+	// when the object no longer matches our composite selector.
+	// This allows the composite to clean up after itself if the object has been
+	// updated to disable the functionality added by the decorator.
+	if pc.finalizeHook.IsEnabled() &&
+		(request.Parent.GetDeletionTimestamp() != nil ||
+			(pc.parentSelector != nil && !pc.parentSelector.Matches(labels.Set(parent.GetLabels())))) {
```

Comment:
nit: It might be better to have this condition split into multiples to improve readability & maintainability

Reply:
done